src/service.js

```diff
@@ -281,38 +281,44 @@ class Service {
 	 * @memberof Service
 	 */
 	_start() {
-		this.logger.debug(`Service '${this.fullName}' is starting...`);
-		return this.Promise.resolve()
-			.then(() => {
-				return this.broker.callMiddlewareHook("serviceStarting", [this]);
-			})
-			.then(() => {
-				// Wait for dependent services
-				if (this.schema.dependencies)
-					return this.waitForServices(
-						this.schema.dependencies,
-						this.settings.$dependencyTimeout || this.broker.options.dependencyTimeout,
-						this.settings.$dependencyInterval || this.broker.options.dependencyInterval
-					);
-			})
-			.then(() => {
-				if (isFunction(this.schema.started))
-					return this.Promise.method(this.schema.started).call(this);
-
-				if (Array.isArray(this.schema.started)) {
-					return this.schema.started
-						.map(fn => this.Promise.method(fn.bind(this)))
-						.reduce((p, fn) => p.then(() => fn()), this.Promise.resolve());
-				}
-			})
-			.then(() => {
-				// Register service
-				return this.broker.registerLocalService(this._serviceSpecification);
-			})
-			.then(() => {
-				return this.broker.callMiddlewareHook("serviceStarted", [this]);
-			})
-			.then(() => this.logger.info(`Service '${this.fullName}' started.`));
+		// create a promise, so even if called multiple times, the `started` handler will only be triggered once
+		if (!this._starting) {
+			this.logger.debug(`Service '${this.fullName}' is starting...`);
+			this._starting = this.Promise.resolve()
+				.then(() => {
+					return this.broker.callMiddlewareHook("serviceStarting", [this]);
+				})
+				.then(() => {
+					// Wait for dependent services
+					if (this.schema.dependencies)
+						return this.waitForServices(
+							this.schema.dependencies,
+							this.settings.$dependencyTimeout ||
+								this.broker.options.dependencyTimeout,
+							this.settings.$dependencyInterval ||
+								this.broker.options.dependencyInterval
+						);
+				})
+				.then(() => {
+					if (isFunction(this.schema.started))
+						return this.Promise.method(this.schema.started).call(this);
+
+					if (Array.isArray(this.schema.started)) {
+						return this.schema.started
+							.map(fn => this.Promise.method(fn.bind(this)))
+							.reduce((p, fn) => p.then(() => fn()), this.Promise.resolve());
+					}
+				})
+				.then(() => {
+					// Register service
+					return this.broker.registerLocalService(this._serviceSpecification);
+				})
+				.then(() => {
+					return this.broker.callMiddlewareHook("serviceStarted", [this]);
+				})
+				.then(() => this.logger.info(`Service '${this.fullName}' started.`));
+		}
+		return this._starting;
```

Comment:
This variable is not nulled after the start process finished.

Reply:
That should be no problem, the resolved promise will always return the same result, also if it would be nulled there's a chance `started` can be called again

Reply:
i removed all changes to service._start / started, the broker changes are sufficient to cover the issue